README.md
# homebridge-mysmartrollershades-bridge

[![mit license](https://badgen.net/badge/license/MIT/red)](https://github.com/apexad/homebridge-mysmartrollershades-bridge/blob/master/LICENSE)
[![npm](https://badgen.net/npm/v/homebridge-mysmartrollershades-bridge)](https://www.npmjs.com/package/homebridge-mysmartrollershades-bridge)
[![npm](https://badgen.net/npm/dt/homebridge-mysmartrollershades-bridge)](https://www.npmjs.com/package/homebridge-mysmartrollershades-bridge)
[![donate](https://badgen.net/badge/donate/paypal/91BE09)](https://www.paypal.me/apexadm)

[Homebridge](https://github.com/homebridge/homebridge) plugin which communicates with MySmartRollerShades through the [MySmartRollerShades/Tilt Bridge](https://www.tiltsmarthome.com/products/smart-hub?variant=31506970116161).  
Configure your roller shades and bridge with the official iOS or Android app first in order to use this Homebridge plugin.

This plugin is not affiliated with the MySmartRollerShades/Tilt Smart Home product

## Features
```diff
- Only adds roller shades to Homebridge/HomeKit! Not yet functional. Work in progress plugin
```

## Configuration
This easiest way to use this plugin is to use [homebridge-config-ui-x](https://www.npmjs.com/package/homebridge-config-ui-x).  
To configure manually, add to the `platforms` section of Homebridge's `config.json` after installing the plugin.

**Config:**
```json
{
  "platform": "MySmartRollerShadesBridge",
  "name": "MySmartRollerShadesBridge",
  "username": "<email address>",
  "password": "<password>"
}
```

Field                   | Description
------------------------|------------
**platform**            | Must always be "MySmartRollerShadesBridge"
**name**                | Best to set to "MySmartRollerShadesBridge"
**username**            | MySmartRollerShades/Tilt app username (usually email address)
**password**            | MySmartRollerShades/Tilt app password
**statusLog**           | _(optional true/false, defaults to false)_ logs position changes
**allowDebug**          | _(optional true/false, defaults to false)_ Outputs a lot of debug info to stdout
**pollingInterval**     | _(optional)_ Polling Interval (in minutes)

## Testing
Use `npx homebridge-mysmartrollershades-bridge` to test your credentials.

## Sponsors
[mbmccormick](https://github.com/mbmccormick)

## Code Credits
[apexad/homebridge-mysmartblinds-bridge](https://github.com/apexad/homebridge-mysmartblinds-bridge) - my own plugin, but check there for other code credits


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.3.0](https://github.com/apexad/homebridge-mysmartrollershades-bridge/compare/v0.2.0...v0.3.0) (2020-12-10)


### Features

* add websocket to settings file ([5973240](https://github.com/apexad/homebridge-mysmartrollershades-bridge/commit/5973240274494de519e3a44393448209e9be77af))
* attempt websocket connection ([de1f83d](https://github.com/apexad/homebridge-mysmartrollershades-bridge/commit/de1f83dcffbda0104ef788f224305f60f343de18))


### Bug Fixes

* add work in progress message to package.json ([95fab92](https://github.com/apexad/homebridge-mysmartrollershades-bridge/commit/95fab92c8cc6267b09cc452f4eb9c33fdbd5d06e))

## [0.2.0](https://github.com/apexad/homebridge-mysmartrollershades-bridge/compare/v0.1.1...v0.2.0) (2020-11-27)


### ⚠ BREAKING CHANGES

* initial creation of roller shades (just shown, not yet controllable)

### Features

* initial creation of roller shades (just shown, not yet controllable) ([a71e63b](https://github.com/apexad/homebridge-mysmartrollershades-bridge/commit/a71e63b5b037c5452a69fd29392aa9a1c7b01663))


### Bug Fixes

* Correct plugin display name ([fdbe90a](https://github.com/apexad/homebridge-mysmartrollershades-bridge/commit/fdbe90a1a3cbed2a304a5c927203ada41b8ce8c5))

### [0.1.1](https://github.com/apexad/homebridge-mysmartrollershades-bridge/compare/v0.1.0...v0.1.1) (2020-11-26)


### Features

* initial homebridge plugin (not yet usable) ([e139b58](https://github.com/apexad/homebridge-mysmartrollershades-bridge/commit/e139b5849b71473c6c50ca3875ef92daf16f0ab7))

## 0.1.0 (2020-11-25)


### Features

* allow npx to test credentials ([9a561bf](https://github.com/apexad/homebridge-mysmartrollershades-bridge/commit/9a561bf6dadf9b22e44e853bdfecb35fe98bc0fd))
* first commit/imported from homebridge-mysmartblinds-bridge/modified test ([fe8fb6c](https://github.com/apexad/homebridge-mysmartrollershades-bridge/commit/fe8fb6ca9c0e7253a0fafcf2762858082b6c1980))


test.js
#!/usr/bin/env node
require("./dist/test.js");

src/test.ts
/* eslint-disable no-console */
import rp from 'request-promise';
import prompt from 'prompt';
import {
  MYSMARTBLINDS_DOMAIN,
  TILTSMARTHOME_OPTIONS,
  TILTSMARTHOME_URL,
} from './settings';

import {
  MySmartRollerShadesAuth,
} from './config';

prompt.get(['username', {name: 'password', hidden: true }], (err: Error, result: MySmartRollerShadesAuth) => {
  rp({
    method: 'POST',
    uri: `https://${MYSMARTBLINDS_DOMAIN}/oauth/token`,
    json: true,
    body: Object.assign(
      {},
      TILTSMARTHOME_OPTIONS,
      {
        username: result.username,
        password: result.password,
      },
    ),
  }).then((response) => {
    console.log(response);
    rp({
      method: 'GET',
      uri: TILTSMARTHOME_URL,
      json: true,
      headers: { Authorization: `Bearer ${response.access_token}` },
    }).then((response) => {
      console.log(JSON.stringify(response));
    });
  });
});


src/platformAccessory.ts
import {
  Service,
  PlatformAccessory,
  CharacteristicValue,
  CharacteristicSetCallback,
} from 'homebridge';
// import rp from 'request-promise';
import { MySmartRollerShadesBridgePlatform } from './platform';

export class MySmartRollerShadesAccessory {
  service!: Service;
  batteryService: Service;
  statusLog: boolean;
  pollingInterval: number;
  name: string;
  id: string;
  platform: MySmartRollerShadesBridgePlatform;
  accessory: PlatformAccessory;
  allowDebug: boolean;

  constructor(
    platform: MySmartRollerShadesBridgePlatform,
    accessory: PlatformAccessory,
  ) {
    this.platform = platform;
    this.name = accessory.context.shade.name;
    this.id = accessory.context.shade.id;
    this.statusLog = platform.config.statusLog || false;
    this.pollingInterval = platform.config.pollingInterval || 0;
    this.allowDebug = platform.config.allowDebug || false;

    accessory.getService(this.platform.Service.AccessoryInformation)!
      .setCharacteristic(this.platform.Characteristic.Manufacturer, 'Tilt Smart Home')
      .setCharacteristic(this.platform.Characteristic.Model, 'MySmartRollerShades')
      .setCharacteristic(this.platform.Characteristic.SerialNumber, this.id);

    this.service = accessory.getService(this.platform.Service.WindowCovering)
    || accessory.addService(this.platform.Service.WindowCovering);

    this.service.setCharacteristic(this.platform.Characteristic.Name, this.name);

    this.service.getCharacteristic(this.platform.Characteristic.TargetPosition)
      .on('set', this.setTargetPosition.bind(this));
    this.updatePosition(accessory.context.shade.shadePosition);

    this.batteryService = accessory.getService(this.platform.Service.BatteryService)
    || accessory.addService(this.platform.Service.BatteryService, `${this.name} Battery`, `${this.id} Battery`);
    this.updateBattery(accessory.context.shade.batteryLevel);
    
    this.accessory = accessory;

    if (this.pollingInterval > 0) {
      if (this.allowDebug) {
        this.platform.log.info(`Begin polling for ${this.name}`);
      }
      setTimeout(() => this.refreshRollerShade(), this.pollingInterval * 1000 * 60);
    }
  }

  updatePosition(currentPosition: number) {
    let reportCurrentPosition = currentPosition;

    if (reportCurrentPosition === 99) {
      reportCurrentPosition = 100;
    }
    if (reportCurrentPosition === 1) {
      reportCurrentPosition = 0;
    }
    if (this.statusLog) {
      this.platform.log.info(`STATUS: ${this.name} updateCurrentPosition : ${reportCurrentPosition} (Actual ${currentPosition})`);
    }

    this.service.updateCharacteristic(this.platform.Characteristic.TargetPosition, reportCurrentPosition);
    this.service.updateCharacteristic(this.platform.Characteristic.CurrentPosition, reportCurrentPosition);
    this.service.updateCharacteristic(this.platform.Characteristic.PositionState, this.platform.Characteristic.PositionState.STOPPED);
  }

  updateBattery(batteryLevel: number) {
    const {
      StatusLowBattery,
    } = this.platform.Characteristic;
    // value of -1 means data was not sent correctly, so ignore it for now

    this.batteryService.updateCharacteristic(this.platform.Characteristic.BatteryLevel, batteryLevel);
    this.batteryService
      .updateCharacteristic(
        StatusLowBattery,
        (batteryLevel < 20 && batteryLevel !== -1) ? StatusLowBattery.BATTERY_LEVEL_LOW : StatusLowBattery.BATTERY_LEVEL_NORMAL,
      );
  }

  setTargetPosition(value: CharacteristicValue, callback: CharacteristicSetCallback) {
    const targetPosition = value as number;
    this.service.updateCharacteristic(this.platform.Characteristic.TargetPosition, targetPosition);

    this.platform.log.info(`${this.name} setTargetPosition to ${value}`);

    /* need to add update logic */
    // update current position
    this.updatePosition(targetPosition);
        
    this.platform.log.info(`${this.name} currentPosition is now ${targetPosition}`);
    callback(null);
  }

  refreshRollerShade() {
    if (this.allowDebug) {
      this.platform.log.info(`Refresh roller shade ${this.name}`);
    }
    /* work in progress
    const shadeState = response.shadeState;
    this.updatePosition(shadeState.position);
    this.updateBattery(shadeState.batteryLevel as number);
    let refreshShadeimeOut = this.pollingInterval * 1000 * 60; // convert minutes to milliseconds
    if (response.headers['x-ratelimit-reset']) {
      refreshShadeTimeOut = new Date(parseInt(response.headers['x-ratelimit-reset']) * 1000).getTime() - new Date().getTime();
      this.platform.log.warn(`Rate Limit reached, refresh for ${this.name} delay to ${new Date(response.headers['x-ratelimit-reset'])}`);
    }
    setTimeout(() => this.refreshRollerShade(), refreshShadeTimeOut);
    */
  }
}


src/platform.ts
import rp from 'request-promise';
import jwt from 'jsonwebtoken';
import WebSocket from 'ws';
import {
  API,
  DynamicPlatformPlugin,
  Logger,
  PlatformAccessory,
  PlatformConfig,
  Service,
  Characteristic,
} from 'homebridge';
import {
  PLATFORM_NAME,
  PLUGIN_NAME,
  MYSMARTBLINDS_DOMAIN,
  TILTSMARTHOME_OPTIONS,
  TILTSMARTHOME_URL,
  TILTSMARTHOME_WSS,
} from './settings';
import {
  MySmartRollerShadesConfig,
  MySmartRollerShadesAuth,
  MySmartRollerShade,
} from './config';
import { MySmartRollerShadesAccessory } from './platformAccessory';

export class MySmartRollerShadesBridgePlatform implements DynamicPlatformPlugin {
  public readonly Service: typeof Service = this.api.hap.Service;
  public readonly Characteristic: typeof Characteristic = this.api.hap.Characteristic;

  // this is used to track restored cached accessories
  public readonly accessories: PlatformAccessory[] = [];
  auth!: MySmartRollerShadesAuth;
  authToken!: string;
  authTokenInterval?: NodeJS.Timeout;
  tiltWebSocket!: WebSocket;
  requestOptions!: {
    method: string;
    uri: string;
    json: boolean;
    headers: {
      Authorization: string;
    };
  };


  constructor(
    public readonly log: Logger,
    public readonly config: PlatformConfig & MySmartRollerShadesConfig,
    public readonly api: API,
  ) {
    /* plugin not configured check */
    if (!config) {
      this.log.info('No configuration found for platform ', PLATFORM_NAME);
      return;
    }

    /* setup config */
    this.config = config;
    this.log = log;

    try {
      if (!this.config.username) {
        throw new Error('MySmartRollerShades Bridge - You must provide a username');
      }
      if (!this.config.password) {
        throw new Error('MySmartRollershades Bridge - You must provide a password');
      }
      this.auth = {
        username: this.config.username,
        password: this.config.password,
      };
    } catch(err) {
      this.log.error(err);
    }

    this.log.debug('Finished initializing platform:', this.config.name);
    this.api.on('didFinishLaunching', () => {
      log.debug('Executed didFinishLaunching callback');
      // run the method to discover / register your devices as accessories
      this.discoverDevices();
    });
  }

  configureAccessory(accessory: PlatformAccessory) {
    this.log.info('Loading roller shade from cache:', accessory.displayName);

    // add the restored accessory to the accessories cache so we can track if it has already been registered
    this.accessories.push(accessory);
  }

  refreshAuthToken() {
    return rp({
      method: 'POST',
      uri: `https://${MYSMARTBLINDS_DOMAIN}/oauth/token`,
      json: true,
      body: Object.assign(
        {},
        TILTSMARTHOME_OPTIONS,
        this.auth,
      ),
    }).then((response) => {
      this.authToken = response.access_token;
      this.requestOptions = {
        method: 'GET',
        uri: TILTSMARTHOME_URL,
        json: true,
        headers: { Authorization: `Bearer ${response.access_token}` },
      };

      if (this.config.allowDebug) {
        const authTokenExpireDate = new Date((jwt.decode(response.id_token || '{ exp: 0 }') as { exp: number }).exp * 1000).toISOString();
        this.log.info(`authToken refresh, now expires ${authTokenExpireDate}`);
      }
    });
  }

  discoverDevices() {
    this.log.info('This plugin is still a work in progress');
    this.refreshAuthToken().then(() => {
      this.authTokenInterval = setInterval(this.refreshAuthToken.bind(this), 1000 * 60 * 60 * 8);
      rp(this.requestOptions).then((response) => {
        // attempt WebSocket setup
        try {
          this.tiltWebSocket = new WebSocket(TILTSMARTHOME_WSS, [], { headers: this.requestOptions.headers });

          this.tiltWebSocket.on('open', () => {
            this.log.warn('WebSocket has connected!');
          });

          this.tiltWebSocket.on('message', (data: string) => {
            this.log.warn(`WebSocket message recieved: ${JSON.parse(data)}`);
          });
        } catch(err) {
          this.log.warn('WebSocket connection failed!', err);
        }

        // create blind accessories
        response.rooms.forEach((room: { rollerShades: MySmartRollerShade[] }) => {
          room.rollerShades.forEach((rollerShade) => {
            const {
              id: shadeID,
              name: shadeName,
            } = rollerShade;
            const uuid = this.api.hap.uuid.generate(shadeID);

            const existingAccessory = this.accessories.find(accessory => accessory.UUID === uuid);
            if (existingAccessory) {
              this.log.debug(`Restore cached roller shade: ${shadeName}`);

              new MySmartRollerShadesAccessory(this, existingAccessory);

              this.api.updatePlatformAccessories([existingAccessory]);
            } else {
              this.log.info(`Adding new roller shade: ${shadeName}`);
        
              // create a new accessory
              const accessory = new this.api.platformAccessory(shadeName, uuid);
              accessory.context.shade = {
                name: shadeName,
                id: shadeID,
                shadePosition: 100, // change to real value via get shade
                batteryLevel: 100, // change to real value via get shade
              };

              new MySmartRollerShadesAccessory(this, accessory);
        
              this.api.registerPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [accessory]);
            }
          });
          // need to figure out how to handle deleted roller shades
        });
      });
    });
  }
}